nixos/modules/services/web-apps/akkoma.nix

```diff
@@ -0,0 +1,1029 @@
+            ref = "stable";
+          };
+        };
+        defaultText = literalExpression ''
+          {
+            primary = {
+              package = pkgs.akkoma-frontends.pleroma-fe;
+              name = "pleroma-fe";
+              ref = "stable";
+            };
+            admin = {
+              package = pkgs.akkoma-frontends.admin-fe;
+              name = "admin-fe";
+              ref = "stable";
+            };
+          }
+        '';
+      };
+
+      dist = {
+        address = mkOption {
+          type = ipAddress;
+          default = "127.0.0.1";
+          description = mdDoc ''
+            Listen address for Erlang distribution protocol and Port Mapper Daemon (epmd).
+          '';
+        };
+
+        epmdPort = mkOption {
+          type = types.port;
+          default = 4369;
+          description = mdDoc "TCP port to bind Erlang Port Mapper Daemon to.";
+        };
+
+        portMin = mkOption {
+          type = types.port;
+          default = 49152;
+          description = mdDoc "Lower bound for Erlang distribution protocol TCP port.";
+        };
+
+        portMax = mkOption {
+          type = types.port;
+          default = 65535;
+          description = mdDoc "Upper bound for Erlang distribution protocol TCP port.";
+        };
+
+        cookie = mkOption {
+          type = types.nullOr secret;
+          default = null;
+          example = { _secret = "/var/lib/secrets/akkoma/releaseCookie"; };
+          description = mdDoc ''
+            Erlang release cookie.
+
+            If set to `null`, a temporary random cookie will be generated.
+          '';
+        };
+      };
+
+      config = mkOption {
+        description = mdDoc ''
+          Configuration for Akkoma. The attributes are serialised to Elixir DSL.
+
+          Refer to <https://docs.akkoma.dev/stable/configuration/cheatsheet/> for
+          configuration options.
+
+          Settings containing secret data should be set to an attribute set containing the
+          attribute `_secret` - a string pointing to a file containing the value the option
+          should be set to.
+        '';
+        type = types.submodule {
+          freeformType = format.type;
+          options = {
+            ":pleroma" = {
+              ":instance" = {
+                name = mkOption {
+                  type = types.nonEmptyStr;
+                  description = mdDoc "Instance name.";
+                };
+
+                email = mkOption {
+                  type = types.nonEmptyStr;
+                  description = mdDoc "Instance administrator email.";
+                };
+
+                description = mkOption {
+                  type = types.nonEmptyStr;
+                  description = mdDoc "Instance description.";
+                };
+
+                static_dir = mkOption {
+                  type = absolutePath;
+                  visible = false;
+                  default = "${cfg.stateDir}/static";
```

Comment:
I think users might appreciate a description of the function of this directory.

I am, even after reading https://docs.pleroma.social/backend/configuration/static_dir/, puzzled whether this directory is only intended to be used as a location to override/add custom static files, or whether akkoma will populate any files there. (If this is only for overrides, consider defaulting to a derivation that builds an empty directory.)

Reply:
The management command can write to this dir. It can be asked to [fetch custom emoji packs](https://docs.akkoma.dev/stable/administration/CLI_tasks/emoji/), in which case it saves them in there. It can also be asked to [fetch frontends](https://docs.akkoma.dev/stable/administration/CLI_tasks/robots_txt/), but since we're doing that via Nix in this module, imperative management is not going to work. 

We could expect users to manage emoji packs via Nix as well, but defaulting to a stateful dir allows for imperative management if people want it. 

As far as I'm aware, Akkoma proper does not write to this dir at runtime.

Reply:
I would suggest in this case starting from a configuration that is more restrictive on how the directory can be used, and making it less restrictive going forward. The reason for that is that going in the opposite direction is much harder without breaking people's configuration.

IIUC emoji are fetched/expected to be present in `$staticDir/emoji/name_of_the_emoji_pack". Are there subdirectories other than `emoji` where you would expect some part of akkoma to wish to write to?

Reply:
There is also a task to generate a `robots.txt`, although that one is arguably less useful

Reply:
I’ll consider creating this directory using a derivation by default, instead of composing it using bind mounts.

Reply:
Do you mean "have a derivation that creates something with a couple of symlinks and copy it over at init time"? If not, how do you intend to accomodate those who wish to install emoji packs?

Reply:
BTW. Note that `systemd.tmpfiles` can be used to ensure existence of symlinks, not only directories.

Reply:
> Do you mean "have a derivation that creates something with a couple of symlinks and copy it over at init time"? If not, how do you intend to accomodate those who wish to install emoji packs?

I’d just set the `static_dir` config option to a directory in the Nix store that contains everything, for example as symlinks, created through a derivation that puts everything together. This could then be overridden by a custom derivation, or pointed to a directory managed by the user as state.

Reply:
Hm~ I don't see why, but having directories in store with symlinks to directories outside of store feels weird to me (IIUC this is how you intend to solve the emoji thing?). I will think whether there's any good reason for that feeling.

Reply:
I wouldn’t want that either. If the user wants emoji packs, we should create derivations for them, or have the user to manage the directory as mutable state. In the latter case the user could for example add symlinks to their preferred frontends in the store.

I could add a derivation for one of the popular emoji packs.

Reply:
Ah, so you want to provide either an option of the directory being totally static, or user-managed?

I'm not entirely sure what you're proposing, but you seem to be having the same constraints in mind that I'd have, so I'll just wait and see the next version.

Adding a derivation for an emoji pack would be very useful, because it would allow us to see how one is supposed to be used and have a working example for users.

Reply:
I added a derivation for one of the standard emoji packs.

The static directory is now generated using a derivation. It relies on `cp --reflink=auto` as symlinks might break confinement and hard links between store paths cannot be created in the derivation itself.

I am not completely satisfied with this solution and a bit unsure about what should be covered by the default derivation. I included terms of service, frontends and emoji packs, but the favicon and robots.txt might be interesting as well.

I feel that a named derivation would be useful so that it can be used as input for custom derivations, but I am still unsure about how to best achieve this. You mentioned using an overlay for this, and for a moment I also considered a read‐only config option.

Reply:
@robryk Could you help me out with the overlay? I fail at defining it properly so that it can be used from the test. Would you be able to provide an example of how you managed to do it?